examples/WeatherForecast/src/WeatherForecast/Program.cs

```diff
@@ -9,9 +12,22 @@
 builder.Services.AddServerSideBlazor();
 builder.Services.AddHttpClient();
 builder.Services.AddFluentUIComponents();
-builder.Services.AddSingleton<IWeatherDataReadOnlyRepository, WeatherDataReadOnlyRepository>();
-builder.Services.AddSingleton<IWeatherDataWriteOnlyRepository, WeatherDataWriteOnlyRepository>();
-builder.Services.AddSingleton<ISearchCityOrZipCode, SearchCityOrZipCode>();
+
+var connectionString = builder.Configuration.GetConnectionString("DefaultConnection");
+
+if (string.IsNullOrWhiteSpace(connectionString))
+{
+  builder.Services.AddSingleton<IWeatherDataReadOnlyRepository, WeatherDataReadOnlyRepository>();
+  builder.Services.AddSingleton<IWeatherDataWriteOnlyRepository, WeatherDataWriteOnlyRepository>();
+  builder.Services.AddSingleton<ISearchCityOrZipCode, SearchCityOrZipCode>();
+}
+else
+{
+  builder.Services.AddDbContext<WeatherDataContext>(options => options.UseSqlServer(connectionString));
+  builder.Services.AddScoped<IWeatherDataReadOnlyRepository, WeatherDataReadOnlyContext>();
+  builder.Services.AddScoped<IWeatherDataWriteOnlyRepository, WeatherDataWriteOnlyContext>();
+  builder.Services.AddScoped<ISearchCityOrZipCode, SearchCityOrZipCode>();
+}
```

Comment:
should we have all singleton? and just

```
if (!string.IsNullOrWhiteSpace(connectionString)) {
    builder.Services.AddDbContext<WeatherDataContext>(options => options.UseSqlServer(connectionString));
}
...
```

Reply:
in general, I wonder if for integration test purposes would have only the else branch. Also, should remove `WeatherDataReadOnlyRepository` and `WeatherDataWriteOnlyRepository` for the example simplicity.

Reply:
I kept the if/else, that devs can run the application without having a database (instead of just running the tests). But we can also use Testcontainers to spin up the database in the application if it is missing, might be a nice gimmick.